core/domain/user_services.py

```diff
@@ -938,14 +1015,61 @@ def create_new_user(gae_id, email):
         raise Exception(
             'User %s already exists for gae_id %s.'
             % (user_settings.user_id, gae_id))
-
     user_id = user_models.UserSettingsModel.get_new_id('')
     user_settings = UserSettings(
         user_id, gae_id, email, feconf.ROLE_ID_EXPLORATION_EDITOR,
         preferred_language_codes=[constants.DEFAULT_LANGUAGE_CODE])
-    _save_user_auth_details(UserAuthDetails(user_id, gae_id))
-    _save_user_settings(user_settings)
-    create_user_contributions(user_id, [], [])
+    user_auth_details = UserAuthDetails(user_id, gae_id)
+    transaction_services.run_in_transaction(
+        _save_new_user_details, user_settings, user_auth_details)
+    return user_settings
+
+
+def create_new_profile(
```

Comment:
You probably should be making a plural version of this (create_new_profiles) -- when an Android app connects to the server for the first time, there'll probably be several profiles on it that need to be uploaded in one go. (Make sure to do a put_multi() at the end.) Then, create_new_profile() can then call create_new_profiles() with just one element in the pin/display_aliases list.

If you think about it that way, then perhaps you may need a domain object to represent ProfileDetails (pin + display alias). 

Also, under what circumstances would it be possible to have a profile without a display alias? I thought display aliases are mandatory for profiles (it's the only way to actually identify them).

And let's say a user is renamed on the Android device while it is offline. When it connects back, how can we tell which user has been renamed? Doesn't there need to be some unchanging profile ID?

Reply:
Profiles can be renamed/created only when connected to the internet. Mentioned this is in the [design doc](https://docs.google.com/document/d/1sQagy3irkjB1Kt5sdb_KkOgxODt2m2ya0ZCoSJNlY4A/edit#bookmark=id.43pnksmjhtye) based on the discussion we had in a separate email thread for syncing of profiles in Android.
--> Given that, I think it's okay to keep the function for creation of one profile only.

> Also, under what circumstances would it be possible to have a profile without a display alias? I thought display aliases are mandatory for profiles (it's the only way to actually identify them).

Yes it is mandatory for them. I am doing that check in `user_settings.validate()` function - if it's a learner role then display alias must be set. (Profiles won't be created until they have a valid `display_alias`).

Reply:
I am slightly surprised by the first answer. @BenHenning could you please confirm?

@sajalasati for the second point, if that's the case, then why are you allowing profile_display_alias=None?

Reply:
That's an implementation mistake, Sorry, I will remove that.

Reply:
Also, just realized I should clarify my thinking for point 1.

There will be a significant group of users who download the app, and then stay offline forever (because they don't have good internet access). Users in this demographic are likely to share a phone between multiple people in a household, hence they will need profiles. They may not care about connecting to the server, at least at the outset. Given this use case, I have concerns about forbidding profiles from being set up when the user is offfline.

Reply:
Yes you are right, and these kind of points did come up during my discussion with @mschanteltc (primarily) while doing the Prod Design. Would it be better to ask this question on that email thread? I have pinged you there.

Reply:
As discussed in the email, let's make plural versions of CRUD operations for profiles

Reply:
Done.